### ISSUE TITLE
Train RL Models to Play Ludo

### ISSUE THREAD

**Issue by KameniAlexNea** [KameniAlexNea/ludo-king-ai#6]
To train an RL agent on your saved Ludo games you’d typically follow these steps:

1. Define your MDP  
  • State space: encode each board position, tokens’ locations, turn order, dice outcome, etc., into a fixed-length vector.  
  • Action space: the set of legal moves (e.g. which token to advance given the dice roll).  
  • Rewards: assign +1 for a win, 0 for non‐terminal steps (or a small shaping reward for capturing an opponent or moving closer to home), and −1 or 0 for losses.

2. Extract transitions from your JSON logs  
  • Parse each saved decision record into tuples (s, a, r, s′). You already have state “ctx”, chosen token “tok” (the action), and “move_res” to infer rewards and next state.  
  • Assemble a replay buffer of these transitions for offline (off‐policy) training or behavioral cloning.

3. Choose an algorithm & framework  
  • For discrete actions you can use DQN or one of its variants, or a policy‐gradient method like PPO/A2C.  
  • Frameworks such as Stable-Baselines3 (Python, PyTorch) or RLlib (Ray) let you plug in your custom Gym-style Ludo environment and replay buffer.

4. Behavioral cloning (optional pre-training)  
  • Train a policy network by supervised learning on (s→a) from your expert games to give the agent a strong initialization.

5. Offline or online RL  
  • Offline: train directly on your logged transitions using an off-policy algorithm (e.g. DQN, Conservative Q-Learning) to bootstrap a policy without further self‐play.  
  • Online: wrap your Ludo game logic in an OpenAI Gym environment, let the agent self-play to collect more data, and train with PPO/A2C or DQN, continually updating its policy.

6. Evaluation and iteration  
  • Periodically pit your agent against baseline strategies (random, heuristic) or holdout expert games to measure win‐rate.  
  • Tune network architecture, reward shaping, and hyperparameters until performance converges.

In short: turn your JSON into a dataset of (state, action, reward, next state) tuples, pick an RL algorithm and framework, optionally pretrain by imitation, then train (offline or via self-play), and evaluate iteratively.

Here is an example of state : 
{
    "timestamp": "2025-08-21T07:00:41.743465",
    "strategy": "Strategy(name=LLM-Ollama, description=Uses ollama cogito model for decisions)",
    "game_context": {
      "current_situation": {
        "player_color": "green",
        "dice_value": 2,
        "consecutive_sixes": 0,
        "turn_count": 5
      },
      "player_state": {
        "player_id": 1,
        "color": "green",
        "start_position": 14,
        "tokens": [
          {
            "token_id": 0,
            "player_color": "green",
            "state": "home",
            "position": -1,
            "is_in_home": true,
            "is_active": false,
            "is_in_home_column": false,
            "is_finished": false
          },
          {
            "token_id": 1,
            "player_color": "green",
            "state": "active",
            "position": 14,
            "is_in_home": false,
            "is_active": true,
            "is_in_home_column": false,
            "is_finished": false
          },
          {
            "token_id": 2,
            "player_color": "green",
            "state": "home",
            "position": -1,
            "is_in_home": true,
            "is_active": false,
            "is_in_home_column": false,
            "is_finished": false
          },
          {
            "token_id": 3,
            "player_color": "green",
            "state": "home",
            "position": -1,
            "is_in_home": true,
            "is_active": false,
            "is_in_home_column": false,
            "is_finished": false
          }
        ],
        "tokens_in_home": 3,
        "active_tokens": 1,
        "tokens_in_home_column": 0,
        "finished_tokens": 0,
        "has_won": false
      },
      "opponents": [
        {
          "color": "red",
          "tokens_finished": 0,
          "tokens_active": 1,
          "threat_level": 0.1
        }
      ],
      "valid_moves": [
        {
          "token_id": 1,
          "current_position": 14,
          "current_state": "active",
          "target_position": 16,
          "move_type": "advance_main_board",
          "is_safe_move": false,
          "captures_opponent": false,
          "strategic_value": 2.0,
          "captured_tokens": []
        }
      ],
      "strategic_analysis": {
        "can_capture": false,
        "can_finish_token": false,
        "can_exit_home": false,
        "safe_moves": [],
        "risky_moves": [
          {
            "token_id": 1,
            "current_position": 14,
            "current_state": "active",
            "target_position": 16,
            "move_type": "advance_main_board",
            "is_safe_move": false,
            "captures_opponent": false,
            "strategic_value": 2.0,
            "captured_tokens": []
          }
        ],
        "best_strategic_move": {
          "token_id": 1,
          "current_position": 14,
          "current_state": "active",
          "target_position": 16,
          "move_type": "advance_main_board",
          "is_safe_move": false,
          "captures_opponent": false,
          "strategic_value": 2.0,
          "captured_tokens": []
        }
      }
    },
    "chosen_move": 1,
    "outcome": {
      "success": true,
      "player_color": "green",
      "token_id": 1,
      "dice_value": 2,
      "old_position": 14,
      "new_position": 16,
      "captured_tokens": [],
      "token_finished": false,
      "extra_turn": false
    }
  }

proposed code from claude : 

Looking at your Ludo game data, you have a rich dataset that's perfect for training a reinforcement learning model! Here's a comprehensive approach to train an RL agent:

## 1. Data Preprocessing

First, let's convert your game data into a suitable format for RL training:

```python
import numpy as np
import json
from collections import deque
import torch
import torch.nn as nn
import torch.nn.functional as F

class LudoStateEncoder:
    def __init__(self, board_size=52, max_tokens=4, num_players=4):
        self.board_size = board_size
        self.max_tokens = max_tokens
        self.num_players = num_players
        self.state_dim = self._calculate_state_dim()
    
    def _calculate_state_dim(self):
        # Token positions (4 tokens * 4 players * board_size+2 for home/finished)
        token_positions = self.num_players * self.max_tokens * (self.board_size + 2)
        # Game context (dice, consecutive_sixes, turn_count, current_player)
        game_context = 4
        # Player statistics (tokens_home, active, finished, won) * players
        player_stats = self.num_players * 4
        return token_positions + game_context + player_stats
    
    def encode_state(self, game_data):
        """Convert game state to numerical vector"""
        state = np.zeros(self.state_dim)
        idx = 0
        
        # Encode token positions for all players
        current_player_id = 0 if game_data['game_context']['current_situation']['player_color'] == 'red' else 1
        
        # Get all players' token states (you'll need to track all players)
        for player_id in range(self.num_players):
            if player_id == current_player_id:
                tokens = game_data['game_context']['player_state']['tokens']
            else:
                # You'll need opponent data - for now, initialize as unknown
                tokens = [{'position': -1, 'state': 'home'} for _ in range(4)]
            
            for token in tokens:
                pos = token.get('position', -1)
                if pos == -1:  # home
                    state[idx] = 1
                elif token.get('is_finished', False):  # finished
                    state[idx + self.board_size + 1] = 1
                else:  # on board
                    state[idx + pos + 1] = 1
                idx += self.board_size + 2
        
        # Encode game context
        context = game_data['game_context']['current_situation']
        state[idx] = context['dice_value'] / 6.0  # normalize
        state[idx + 1] = context['consecutive_sixes'] / 3.0  # max 3 sixes
        state[idx + 2] = min(context['turn_count'], 100) / 100.0  # normalize turn count
        state[idx + 3] = current_player_id / (self.num_players - 1)  # current player
        idx += 4
        
        # Encode player statistics
        player_state = game_data['game_context']['player_state']
        state[idx] = player_state['tokens_in_home'] / 4.0
        state[idx + 1] = player_state['active_tokens'] / 4.0
        state[idx + 2] = player_state['finished_tokens'] / 4.0
        state[idx + 3] = 1.0 if player_state['has_won'] else 0.0
        
        return state
    
    def encode_action(self, chosen_move, valid_moves):
        """Convert chosen move to action index"""
        if chosen_move < len(valid_moves):
            return chosen_move
        return 0  # default action
```

## 2. Deep Q-Network (DQN) Implementation

```python
class LudoDQN(nn.Module):
    def __init__(self, state_dim, action_dim=4, hidden_dim=512):
        super(LudoDQN, self).__init__()
        self.state_dim = state_dim
        self.action_dim = action_dim
        
        self.network = nn.Sequential(
            nn.Linear(state_dim, hidden_dim),
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(hidden_dim, hidden_dim),
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(hidden_dim, hidden_dim // 2),
            nn.ReLU(),
            nn.Linear(hidden_dim // 2, action_dim)
        )
    
    def forward(self, state):
        return self.network(state)

class LudoDQNAgent:
    def __init__(self, state_dim, action_dim=4, lr=0.001, gamma=0.99, 
                 epsilon=1.0, epsilon_decay=0.995, epsilon_min=0.01):
        self.state_dim = state_dim
        self.action_dim = action_dim
        self.gamma = gamma
        self.epsilon = epsilon
        self.epsilon_decay = epsilon_decay
        self.epsilon_min = epsilon_min
        
        # Neural networks
        self.q_network = LudoDQN(state_dim, action_dim)
        self.target_network = LudoDQN(state_dim, action_dim)
        self.optimizer = torch.optim.Adam(self.q_network.parameters(), lr=lr)
        
        # Experience replay
        self.memory = deque(maxlen=10000)
        self.batch_size = 32
        
        # Update target network
        self.update_target_network()
    
    def update_target_network(self):
        self.target_network.load_state_dict(self.q_network.state_dict())
    
    def remember(self, state, action, reward, next_state, done):
        self.memory.append((state, action, reward, next_state, done))
    
    def act(self, state, valid_moves):
        """Choose action using epsilon-greedy policy with valid move masking"""
        if np.random.random() <= self.epsilon:
            return np.random.choice(len(valid_moves))
        
        state_tensor = torch.FloatTensor(state).unsqueeze(0)
        q_values = self.q_network(state_tensor)
        
        # Mask invalid actions
        masked_q_values = q_values.clone()
        valid_indices = list(range(len(valid_moves)))
        
        if len(valid_indices) < self.action_dim:
            # Set invalid actions to very low values
            for i in range(self.action_dim):
                if i not in valid_indices:
                    masked_q_values[0][i] = -float('inf')
        
        return masked_q_values.argmax().item() if len(valid_indices) > 0 else 0
    
    def replay(self):
        if len(self.memory) < self.batch_size:
            return
        
        batch = np.random.choice(len(self.memory), self.batch_size, replace=False)
        states = torch.FloatTensor([self.memory[i][0] for i in batch])
        actions = torch.LongTensor([self.memory[i][1] for i in batch])
        rewards = torch.FloatTensor([self.memory[i][2] for i in batch])
        next_states = torch.FloatTensor([self.memory[i][3] for i in batch])
        dones = torch.BoolTensor([self.memory[i][4] for i in batch])
        
        current_q_values = self.q_network(states).gather(1, actions.unsqueeze(1))
        next_q_values = self.target_network(next_states).max(1)[0].detach()
        target_q_values = rewards + (self.gamma * next_q_values * ~dones)
        
        loss = F.mse_loss(current_q_values.squeeze(), target_q_values)
        
        self.optimizer.zero_grad()
        loss.backward()
        self.optimizer.step()
        
        if self.epsilon > self.epsilon_min:
            self.epsilon *= self.epsilon_decay
```

## 3. Training Pipeline

```python
class LudoRLTrainer:
    def __init__(self, game_data_file):
        self.encoder = LudoStateEncoder()
        self.agent = LudoDQNAgent(self.encoder.state_dim)
        self.game_data = self.load_game_data(game_data_file)
        
    def load_game_data(self, file_path):
        with open(file_path, 'r') as f:
            return json.load(f)
    
    def calculate_reward(self, game_data, next_game_data=None):
        """Calculate reward based on game outcome and strategic value"""
        outcome = game_data.get('outcome', {})
        context = game_data['game_context']
        
        reward = 0.0
        
        # Basic move rewards
        if outcome.get('success', False):
            reward += 1.0
        
        # Strategic rewards
        if outcome.get('captured_tokens', []):
            reward += 10.0 * len(outcome['captured_tokens'])
        
        if outcome.get('token_finished', False):
            reward += 20.0
        
        if outcome.get('extra_turn', False):
            reward += 5.0
        
        # Use pre-calculated strategic value
        strategic_analysis = context.get('strategic_analysis', {})
        best_move = strategic_analysis.get('best_strategic_move', {})
        chosen_move_idx = game_data.get('chosen_move', 0)
        valid_moves = context.get('valid_moves', [])
        
        if chosen_move_idx < len(valid_moves):
            chosen_move = valid_moves[chosen_move_idx]
            reward += chosen_move.get('strategic_value', 0) * 0.1
        
        # Win condition
        if context['player_state'].get('has_won', False):
            reward += 100.0
        
        return reward
    
    def create_training_sequences(self):
        """Convert game data to training sequences"""
        sequences = []
        current_sequence = []
        
        for i, game_data in enumerate(self.game_data):
            state = self.encoder.encode_state(game_data)
            action = self.encoder.encode_action(
                game_data['chosen_move'], 
                game_data['game_context']['valid_moves']
            )
            reward = self.calculate_reward(game_data)
            
            if i < len(self.game_data) - 1:
                next_state = self.encoder.encode_state(self.game_data[i + 1])
                done = False
            else:
                next_state = state  # terminal state
                done = True
            
            current_sequence.append((state, action, reward, next_state, done))
            
            # End sequence when game ends or player changes
            if done or (i < len(self.game_data) - 1 and 
                       game_data['game_context']['current_situation']['player_color'] != 
                       self.game_data[i + 1]['game_context']['current_situation']['player_color']):
                sequences.append(current_sequence)
                current_sequence = []
        
        return sequences
    
    def train(self, epochs=1000, target_update_freq=100):
        """Train the RL agent"""
        sequences = self.create_training_sequences()
        
        print(f"Training on {len(sequences)} game sequences...")
        
        for epoch in range(epochs):
            total_loss = 0
            
            # Sample sequences for training
            for sequence in sequences:
                for state, action, reward, next_state, done in sequence:
                    self.agent.remember(state, action, reward, next_state, done)
            
            # Train the agent
            if len(self.agent.memory) > self.agent.batch_size:
                self.agent.replay()
            
            # Update target network periodically
            if epoch % target_update_freq == 0:
                self.agent.update_target_network()
            
            if epoch % 100 == 0:
                print(f"Epoch {epoch}, Epsilon: {self.agent.epsilon:.3f}")
        
        print("Training completed!")
    
    def save_model(self, path):
        torch.save(self.agent.q_network.state_dict(), path)
    
    def load_model(self, path):
        self.agent.q_network.load_state_dict(torch.load(path))
        self.agent.target_network.load_state_dict(torch.load(path))
```

## 4. Integration with Your Game

```python
class LudoRLPlayer:
    def __init__(self, model_path=None):
        self.encoder = LudoStateEncoder()
        self.agent = LudoDQNAgent(self.encoder.state_dim)
        
        if model_path:
            self.agent.load_model(model_path)
            self.agent.epsilon = 0.0  # No exploration during play
    
    def choose_move(self, game_context):
        """Choose move for the RL agent"""
        # Convert your game context to the expected format
        game_data = {
            'game_context': game_context,
            'chosen_move': 0  # placeholder
        }
        
        state = self.encoder.encode_state(game_data)
        valid_moves = game_context.get('valid_moves', [])
        
        if not valid_moves:
            return 0
        
        action = self.agent.act(state, valid_moves)
        return min(action, len(valid_moves) - 1)
```

## 5. Usage Example

```python
# Training
trainer = LudoRLTrainer('your_game_data.json')
trainer.train(epochs=2000)
trainer.save_model('ludo_dqn_model.pth')

# Playing against human
rl_player = LudoRLPlayer('ludo_dqn_model.pth')

# In your game loop:
def get_ai_move(game_context):
    return rl_player.choose_move(game_context)
```

## 6. Advanced Improvements

Consider these enhancements:

1. **Policy Gradient Methods**: Implement PPO or A3C for better performance
2. **Multi-Agent Training**: Train multiple agents against each other
3. **Curriculum Learning**: Start with simpler scenarios and gradually increase complexity
4. **Feature Engineering**: Add more sophisticated state representations
5. **Self-Play**: Let the agent play against itself to improve

This framework should give you a solid foundation for training an RL agent to play Ludo against humans. The key is to iterate and refine based on the agent's performance!
